### PR TITLE
[ConstitutiveLawsApp] Adding a consistent Plastic-Damage model and minor improvements

### PR DESCRIPTION
**Pull request by AlejandroCornejo**
In this PR I'm adding a new Plastic-Damage model. I'll add a test before merging.

Additionally I've added another numerical tangent estimation and corrected a bug in another one (first order was not working)

I know there are a lot of commits but only inside the CL :) (tough implementation and testing)

Adding a test ASAP (before merging) for the tangent calculation and for the CL

### CODE REVIEW

File: applications/ConstitutiveLawsApplication/constitutive_laws_application.cpp [hunk 1]
```diff
@@ -443,6 +443,10 @@ void KratosConstitutiveLawsApplication::Register()
     KRATOS_REGISTER_CONSTITUTIVE_LAW("ParallelRuleOfMixturesLaw2D", mParallelRuleOfMixturesLaw2D);
     KRATOS_REGISTER_CONSTITUTIVE_LAW("ParallelRuleOfMixturesLaw3D", mParallelRuleOfMixturesLaw3D);
 
+    KRATOS_REGISTER_CONSTITUTIVE_LAW("AssociativePlasticDamageModel3DVonMisesVonMises", mAssociativePlasticDamageModel3DVonMisesVonMises);
```
Comment (loumalouomega) [inline applications/ConstitutiveLawsApplication/constitutive_laws_application.cpp:447]: imagen

File: applications/ConstitutiveLawsApplication/custom_constitutive/associative_plastic_damage_model.cpp [hunk 2]
```diff
@@ -0,0 +1,1092 @@
...
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<class TYieldSurfaceType>
+void AssociativePlasticDamageModel<TYieldSurfaceType>::CalculateElasticComplianceMatrix(
+    BoundedMatrixType& rC,
+    const Properties& rMaterialProperties
+    )
+{
+    const double E = rMaterialProperties[YOUNG_MODULUS];
```
Comment (loumalouomega) [inline applications/ConstitutiveLawsApplication/custom_constitutive/associative_plastic_damage_model.cpp:495]: Is this the isotropic constitutive tensor? (should be in utilities to reduce code duplciation, just saying...)

**Review by loumalouomega — CHANGES_REQUESTED** [review]
I do not understand why creating new laws that replicate a high percentage of code of the existing code. Don't you prefer to add the analytics ones directly with the corresponding ifs?

File: applications/ConstitutiveLawsApplication/custom_utilities/tangent_operator_calculator_utility.h [hunk 3]
```diff
@@ -207,6 +207,42 @@ class TangentOperatorCalculatorUtility
                 // Finally we compute the components
                 CalculateComponentsToTangentTensorSecondOrder(auxiliar_tensor, strain_plus, strain_minus, stress_plus, stress_minus, i_component);
 
+                // Reset the values to the initial ones
+                noalias(r_perturbed_strain) = unperturbed_strain_vector_gp;
+                noalias(r_perturbed_integrated_stress) = unperturbed_stress_vector_gp;
+            }
+        } else if (ApproximationOrder == 4) {
```
Comment (loumalouomega) [inline applications/ConstitutiveLawsApplication/custom_utilities/tangent_operator_calculator_utility.h]: 4th order?

File: applications/StructuralMechanicsApplication/structural_mechanics_application_variables.h [hunk 4]
```diff
@@ -38,7 +38,7 @@ namespace Kratos
 
     enum class SofteningType {Linear = 0, Exponential = 1, HardeningDamage = 2};
 
-    enum class TangentOperatorEstimation {Analytic = 0, FirstOrderPerturbation = 1, SecondOrderPerturbation = 2, Secant = 3};
+    enum class TangentOperatorEstimation {Analytic = 0, FirstOrderPerturbation = 1, SecondOrderPerturbation = 2, Secant = 3, SecondOrderPerturbationV2 = 4};
```
Comment (loumalouomega) [inline applications/StructuralMechanicsApplication/structural_mechanics_application_variables.h]: Ok, is not 4th order, is SecondOrderPerturbationV2. In what consists?, can you add a comment so we are not confused?

(did you add the test?)

File: applications/ConstitutiveLawsApplication/custom_constitutive/associative_plastic_damage_model.cpp (hunk 2 above)
Comment (AlejandroCornejo) [reply applications/ConstitutiveLawsApplication/custom_constitutive/associative_plastic_damage_model.cpp:495]: Not exactly Vicente! this methos computes the Compliance Elastic matrix (inverse of the constitutive matrix), I think that this is only used in here

File: applications/ConstitutiveLawsApplication/custom_utilities/tangent_operator_calculator_utility.h (hunk 3 above)
Comment (AlejandroCornejo) [reply applications/ConstitutiveLawsApplication/custom_utilities/tangent_operator_calculator_utility.h]: Nope, in the enum is second order but with another approach. It is computed as first order but taking into account the second derivatives... (really close to central differences)

Comment (loumalouomega) [reply applications/ConstitutiveLawsApplication/custom_utilities/tangent_operator_calculator_utility.h]:
```suggestion
        } else if (ApproximationOrder == 4) { // Second order but with another approach. It is computed as first order but taking into account the second derivatives
```

File: applications/ConstitutiveLawsApplication/custom_constitutive/associative_plastic_damage_model.cpp (hunk 2 above)
Comment (loumalouomega) [reply applications/ConstitutiveLawsApplication/custom_constitutive/associative_plastic_damage_model.cpp:495]: OK, just to know

File: applications/StructuralMechanicsApplication/structural_mechanics_application_variables.h (hunk 4 above)
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/structural_mechanics_application_variables.h]: I'll add a test, indeed

File: applications/StructuralMechanicsApplication/structural_mechanics_application_variables.h [hunk 5]
```diff
@@ -339,6 +339,8 @@ namespace Kratos
     KRATOS_DEFINE_APPLICATION_VARIABLE(STRUCTURAL_MECHANICS_APPLICATION, Vector, STRESS_ON_NODE );
     KRATOS_DEFINE_APPLICATION_VARIABLE(STRUCTURAL_MECHANICS_APPLICATION, std::string, DESIGN_VARIABLE_NAME);
     KRATOS_DEFINE_APPLICATION_VARIABLE(STRUCTURAL_MECHANICS_APPLICATION, double, SERIAL_PARALLEL_EQUILIBRIUM_TOLERANCE);
+    KRATOS_DEFINE_APPLICATION_VARIABLE(STRUCTURAL_MECHANICS_APPLICATION, double, PLASTIC_DAMAGE_PROPORTION);
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/structural_mechanics_application_variables.h]: Please move it to claw if it is not used in structural

File: applications/StructuralMechanicsApplication/structural_mechanics_application_variables.h (hunk 4 above)
Comment (philbucher) [inline applications/StructuralMechanicsApplication/structural_mechanics_application_variables.h]: This is no longer used in structural right?
Can you move it to claw?

**Review by philbucher — CHANGES_REQUESTED** [review]

File: applications/StructuralMechanicsApplication/structural_mechanics_application_variables.h (hunk 4 above)
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/structural_mechanics_application_variables.h]: I've moved ALL the variables only used in the CL :)

File: applications/StructuralMechanicsApplication/structural_mechanics_application_variables.h (hunk 5 above)
Comment (AlejandroCornejo) [reply applications/StructuralMechanicsApplication/structural_mechanics_application_variables.h]: All of them moved to the CLapp

**Review by philbucher — DISMISSED** [review]
:+1: :+1: :+1:

**Comment by AlejandroCornejo** [comment]
I'll wait to add the test to https://github.com/KratosMultiphysics/Kratos/pull/8650#pullrequestreview-639649730 to be merged

**Comment by AlejandroCornejo** [comment]
I've added a test in python in which I'm am **checking the behaviour of the new law AND using the new tangent approximation** (2 tests in one :D).
 Also corrected some variables that have to be inside the SMA

**Review by philbucher — COMMENTED** [review]
@AlejandroCornejo I checked, those variables are only used in some tests, can you please move them back?
=> all those tests already have the required `skiptests`, you only need to change where they come from

**Comment by AlejandroCornejo** [comment]
> 
> 
> @AlejandroCornejo I checked, those variables are only used in some tests, can you please move them back?
> => all those tests already have the required `skiptests`, you only need to change where they come from

Hi! you mean moving the test to the new app?

**Comment by AlejandroCornejo** [comment]
Now the variables are back to the CLApp and I've updated the tests. Now it should be fine @philbucher

**Comment by philbucher** [comment]
> Now the variables are back to the CLApp and I've updated the tests. Now it should be fine @philbucher

perfect thx thats what I meant

**Comment by AlejandroCornejo** [comment]
@loumalouomega can you approve now? 👯‍♂️

**Review by loumalouomega — APPROVED** [review]

**Pull request merged**